popup_offer/src/Plugin/Commerce/PromotionOffer/PopupOffer.php

```diff
@@ -1,293 +0,0 @@
-        '#type' => 'entity_autocomplete',
-        '#target_type' => 'commerce_product',
-    ];
-    $form['discount']['product2'] = [
-        '#type' => 'entity_autocomplete',
-        '#title' => $this->t('Name of discrount product'),
-        '#default_value' => $this->getProductNameByID([$this->configuration['discount_product2']]),
-        '#type' => 'entity_autocomplete',
-        '#target_type' => 'commerce_product',
-    ];
-    $form['discount']['product3'] = [
-        '#type' => 'entity_autocomplete',
-        '#title' => $this->t('Name of discount product'),
-        '#default_value' => $this->getProductNameByID([$this->configuration['discount_product3']]),
-        '#type' => 'entity_autocomplete',
-        '#target_type' => 'commerce_product',
-    ];
-    $form['discount']['product4'] = [
-        '#type' => 'entity_autocomplete',
-        '#title' => $this->t('Name of discount product'),
-        '#default_value' => $this->getProductNameByID([$this->configuration['discount_product4']]),
-        '#type' => 'entity_autocomplete',
-        '#target_type' => 'commerce_product',
-    ];
-    $form['discount']['value'] = [
-        '#type' => 'commerce_number',
-        '#title' => $this->t('Enter the value of discount in % from 0 to 100'),
-        '#default_value' => Calculator::multiply((string)$this->configuration['discount_value'], 100),
-    ];
-
-    return $form;
-  }
-
-
-  /**
-   * {@inheritdoc}
-   */
-  public function validateConfigurationForm(array &$form, FormStateInterface $form_state)
-  {
-    parent::validateConfigurationForm($form, $form_state);
-    $values = $form_state->getValue($form['#parents']);
-    if (!is_numeric($values['product']['name']) || !is_numeric($values['product']['name'])) {
-      $form_state->setError($form, $this->t('You should specify target product and discounts.'));
-    }
-    if (!is_numeric($values['discount']['value'])) {
-      $form_state->setError($form, $this->t('Please specify discount value from 0 to 100'));
-    }
-  }
-
-  /**
-   * {@inheritdoc}
-   */
-  public function submitConfigurationForm(array &$form, FormStateInterface $form_state)
-  {
-
-    parent::submitConfigurationForm($form, $form_state);
-
-    if (!$form_state->getErrors()) {
-      $values = $form_state->getValue($form['#parents']);
-      $this->configuration['main_product'] = $values['product']['name'];
-      $this->configuration['discount_product1'] = $values['discount']['product1'];
-      $this->configuration['discount_product2'] = $values['discount']['product2'];
-      $this->configuration['discount_product3'] = $values['discount']['product3'];
-      $this->configuration['discount_product4'] = $values['discount']['product4'];
-      $this->configuration['discount_value'] = Calculator::divide((string)$values['discount']['value'], '100');
-
-    }
-  }
-
-  /**
-   * {@inheritdoc}
-   */
-  public function apply(EntityInterface $entity, PromotionInterface $promotion, CouponInterface $coupon = NULL)
```

Comment:
it seems you don't check promotions conditions so it's almost unusable because will be applied every time

Reply:
In apply function I am taking variant id of current ordered product and saved variants ids of target product which saved in config
if (in_array($variantId, $configs['variations'])) {
  // If current $variantId within available for this product variatntIds do something....
}